design/view-change.md

```diff
@@ -0,0 +1,229 @@
+  _(i, vd)_ for each view-change certificate in _S_, and _X_ is a structure
+  containing selected _(ck, cd)_ and list of requests to get into new view.
+  
+  Backups consider _NEW-VIEW_ message valid when they have correct matching
+  _VIEW-CHANGE_ message for each entry in _V_ and agrees on contents of _X_.
+  If backup finds that received _NEW-VIEW_ message as invalid it immediately
+  enters view _v+2_, repeating all previous steps.
+  
+  There was a parapgraph in original papers stating that node that sent
+  _VIEW-CHANGE_ message could be uncooperative and don't send it to some
+  nodes. In that case since primary has a view-change certificate it is
+  guaranteed that it is possible to get required _VIEW-CHANGE_ from primary
+  and _f_ matching _VIEW-CHANGE-ACKS_ from non-faulty nodes. Most papers from
+  original authors don't describe exact protocol for this case, but in
+  [Miguel Castro master thesis](https://www.microsoft.com/en-us/research/wp-content/uploads/2017/01/thesis-mcastro.pdf) 
+  there is section 5.2 on solving problem of eventually delivering all 
+  messages even if underlying network can randomly drop them. Their proposed 
+  solution is to periodically broadcast status messages, so other nodes can 
+  learn about missing messages and retransmit them.
+
+- After obtaining (or generating in case of new primary) correct _NEW-VIEW_
+  message each replica sets stable checkpoint to _(ck, cd)_ from _X_ and
+  records all requests in _X_ as pre-prepared. Backup replicas also broadcast
+  _PREPARE_ messages for these requests. Since preparing (by primary) and
+  checking (by backups) of _X_ includes update of uncommited state and
+  dynamic validation of requests it is guaranteed that at this point all
+  correct replicas will have same uncommited state corresponding to
+  pre-prepared requests that got to new view. After that processing goes on
+  in normal state. There are some caveats however.
+
+  First one (also mentioned by authors of PBFT) is that it can happen so 
+  that some replica won't have checkpoint _(ck, cd)_ and it would require 
+  state transfer protocol to get to correct state. This could be done 
+  independent from normal request processing (they can still get prepare and 
+  commit certificates), which will be executed when state is caught up.
+  There is already similar protocol implemented in indy plenum called 
+  catch up, hopefully it could be used as is in this situation.
+  
+  Another problem is that we need somehow to skip execution of requests that
+  were ordered in previous view but view-change protocol started their
+  processing again. Authors of PBFT rely on uniqueness of ppSeqNo of requests,
+  so if node gets commit certificate for some request that it already executed
+  it just skips it. However indy plenum implementation currently resets 
+  ppSeqNo counter upon entering new view, so it cannot be relied upon.
+  One possible workaround is to check if request was already ordered using
```

Comment:
We have a mapping to check if a request has been written to the ledger or not, it is stored in `seqNoDB` of `Node`. It stores the map of `sha256(identifier||req_id) -> txn seq_no `